nova_vm/src/ecmascript/execution/agent.rs

```diff
@@ -52,6 +52,31 @@ pub trait HostHooks: std::fmt::Debug {
     fn host_has_source_text_available(&self, func: Function) -> bool;
 }
 
+pub struct BoxedAgent {
+    agent: Box<Agent>,
+    root_realms: Vec<RealmIdentifier>,
+}
+
+impl BoxedAgent {
+    pub fn new(options: Options, host_hooks: &'static dyn HostHooks) -> Self {
+        Self {
+            agent: Box::new(Agent::new(options, host_hooks)),
+            root_realms: Vec::with_capacity(1),
+        }
+    }
+
+    pub fn with<'agent, F, R>(&'agent mut self, func: F) -> R
+    where
+        F: FnOnce(&'agent mut Agent, &'agent mut Vec<RealmIdentifier>) -> R,
```

Comment:
**issue:** This signature allows returning the inner `&mut Agent` reference, or values with a derived lifetime, out of the closure. This would not be unsound at this point, since the lifetime of those references is tied to the lifetime of the `BoxedAgent`, and they all must be dropped before calling `gc()`. But it might be relevant for future extensions.

You can make the closure's trait bound [higher-ranked](https://doc.rust-lang.org/nomicon/hrtb.html) to fix this, since any choice of `'agent` would last for the duration of the closure call but might not remain alive afterwards:
```suggestion
    pub fn with<F, R>(&mut self, func: F) -> R
    where
        F: for<'agent> FnOnce(&'agent mut Agent, &'agent mut Vec<RealmIdentifier>) -> R,
```

Reply:
Ah, thank you <3